581 notes (#978)
Dyalog/documentation: language-reference-guide/docs/primitive-functions/notes.md

## Implicit Arguments

_Implicit arguments_ are system functions that affect the result of a primitive function. They are all variables that have a global effect, only limited by the current namespace and function scope (if localised). Implicit argument propagate to newly created namespaces and defined functions at creation/definition time.

!!! Hint "Hints and Recommendations"

>     Unexpected values of implicit arguments, especially `⎕IO` and `⎕ML`, are a frequent source of confusion. Keep deviations from the defaults limited and local. Check the implicit arguments if a primitive seems broken. The status bar of Ride and the Windows IDE highlight in red when the current value differs from the default as set by [DEFAULT_IO](../../windows-installation-and-configuration-guide/configuration-parameters/default-io.md) and [DEFAULT_ML](../../windows-installation-and-configuration-guide/configuration-parameters/default-ml.md). 

Reviewer: ```suggestion
    Unexpected values of implicit arguments, especially `⎕IO` and `⎕ML`, are a frequent source of confusion. Dyalog Ltd strongly recommends ensuring that deviations from the default values are limited and local. If a primitive function is not giving the expected result, check the values of its implicit arguments. The status bar of Ride and the Windows IDE shows the current value of `⎕IO` and `⎕ML` and highlights them if they differ from the default values as set by [DEFAULT_IO](../../windows-installation-and-configuration-guide/configuration-parameters/default-io.md) and [DEFAULT_ML](../../windows-installation-and-configuration-guide/configuration-parameters/default-ml.md).
```

      1 2 + 1 2 3         ⍝ same rank, different length
LENGTH ERROR: Mismatched left and right argument shapes
      1 2+1 2 3
         ∧
```

## Scalar and Singleton Extension

A _singleton_ is an array with exactly one element, of any rank: a scalar, a one-element vector, a one-by-one matrix, and so on. An array `Y` is a singleton if `1=×/⍴Y`. When a dyadic scalar function is applied to a singleton and a non-singleton, the singleton is extended to conform with the other argument, and the result takes the shape of the non-singleton argument.

<h2 class="example">Examples</h2>
```apl
      (1 1⍴5) + 2 3⍴⍳6    ⍝ a one-by-one matrix extends to the matrix
6  7  8
9 10 11
      (,5) + 2 3⍴⍳6       ⍝ a one-element vector likewise
6  7  8
9 10 11
```

If both arguments are singletons, the result takes the shape of the one with the higher rank.
```apl
      ⍴ (1 1⍴5) + ,2
1 1
```

An array with a one-length axis but more than one element (such as a `1 3` matrix) is not a singleton and does not extend.
```apl
      (1 3⍴1 2 3) + 2 3⍴⍳6
LENGTH ERROR: Mismatched left and right argument shapes
      (1 3⍴1 2 3)+2 3⍴⍳6
                 ∧
```

!!! Hint "Hints and Recommendations"

>     Singleton extension can lead to surprising or inconsistent results, and it is therefore best to rely on scalar extension only. In this example, `N←1` returns a matrix while all other values return a vector:

Reviewer: ```suggestion
    Singleton extension can lead to surprising or inconsistent results; it is, therefore, recommended to rely on scalar extension only. In this example, `N←1` returns a matrix while all other values return a vector:
```

    ```apl
          ⍴(1 1⍴10)+⍳N←3
    3
          ⍴(1 1⍴10)+⍳N←2
    2
          ⍴(1 1⍴10)+⍳N←1
    1 1
          ⍴(1 1⍴10)+⍳N←0
    0
    ```

## Fill Elements and Prototypes

> Some primitive functions can include _fill elements_ in their result. The fill element for an array is the enclosed first _prototypical element_ or _prototype_ of the array. The _type_ function ([`∊`](type.md) with `⎕ML←0`) converts an array to consist entirely of prototypical elements; `⎕NULL`s stay, numbers become zeros, character become spaces, and instances are replaced with new instances of their classes (if they have niladic constructors). With `⎕ML←1` (the default), _type_ can be written as `⊃0⍴⊂` which exploits the prototype, since, if the array is empty, `⊃Y` gives the disclosed prototype of `Y`.

Reviewer: ```suggestion
Some primitive functions can include _fill elements_ in their result. The fill element for an array is the enclosed first _prototypical element_ or _prototype_ of the array. The _type_ function ([`∊`](type.md) with `⎕ML←0`) converts an array to consist entirely of prototypical elements; `⎕NULL`s remain unchanged, numbers become zeros, character become spaces, and instances are replaced with new instances of their classes (if they have niladic constructors). With `⎕ML←1` (the default), _type_ can be written as `⊃0⍴⊂` (this exploits the prototype, since, if the array is empty, `⊃Y` gives the disclosed prototype of `Y`).
```

    │0 0 0│0 0 0│0 0 0│
    └─────┴─────┴─────┘
    ```

## Axis Specification

Axis specification can be applied to certain mixed primitive functions and to all [dyadic scalar functions](primitive-functions-by-category/#dyadic-scalar-functions). An integer axis identifies a specific axis along which the function is to be applied to one or both of its arguments. Mixed primitive functions have a default axis, either the first or last.

<h2 class="example">Example</h2>
```apl
      1 0 1 ⌿[2] 2 3⍴⍳6
1 3
4 6

      1 2 3 +[2] 2 3⍴10 20 30
11 22 33
11 22 33
```

Some mixed primitive functions allow a fractional axis value, indicating that a new axis or axes are to be created between the axes identified by the lower and upper integer bounds of the value (either of which might not exist).

<h2 class="example">Example</h2>

```apl
      'NAMES' ,[0.5] '='
NAMES
=====
```

Some mixed primitive functions allow a vector axis value, indicating an operation over a collection of axes.

<h2 class="example">Example</h2>

```apl
      ⍴ ,[2 3] 1 3 4 2⍴⎕A
1 12 2
```

`⎕IO` is an [implicit argument](#implicit-arguments) of axis specification.

!!! Hint "Hints and Recommendations"

>     Axis specification only applies to a specific set of primitives and never to defined functions. The [_rank operator_](../primitive-operators/rank.md) is a more general mechanism. If the function operates along the leading axis, then it can substitute for most axis specifications, espcially in combination with [`⍉`](transpose.md). For example:

Reviewer: ```suggestion
    Axis specification only applies to a specific set of primitives and never to defined functions. The [_rank_ operator](../primitive-operators/rank.md) is a more general mechanism. If the function operates along the leading axis, then it can substitute for most axis specifications, particularly in combination with [`⍉`](transpose.md). For example:
```

    ```apl
          1 0 1(⌿⍤1) 2 3⍴⍳6
    1 3
    4 6
          'NAMES' (⍉,⍤0) '='
    NAMES
    =====
    ```

>     If we wanted to replace `,` with a defined function that does something similar, then the surrounding code with be identical for the rank operator whereas axis specification would fail:

Reviewer: ```suggestion
    If we wanted to replace `,` with a defined function that does something similar, then the surrounding code would be identical for the _rank_ operator but axis specification would fail:
```
